autoextract/aio/retry.py

```diff
@@ -104,6 +123,8 @@ def __init__(self):
         self.stop_on_throttling_error = stop_never
         self.stop_on_network_error = stop_after_delay(15 * 60)
         self.stop_on_server_error = self.stop_on_network_error
+        self.stop_on_retriable_query_error = self.stop_on_network_error
+        self.stop_on_domain_occupied_query_error = stop_never
```

Comment:
If I'm not mistaken, this can be problematic.

For example, you create a file with 100K URLs to website A, and 100K URLs to many other websites. Then you start a CLI tool, with 40 connections to the API. Over time you start getting "domain occupied" messages for the website A. This all happens inside request_raw function, so if a domain occupied message is received, the connection won't be reused until the retrials are finished. So in this example all processing stops, all connections are busy retrying domain occupied errors for website A, and work for other websites is not being done.

It can be solved e.g. by having a separate component which limits an amount of requests sent in parallel to a domain. But this is not straightforward to implement, and we don't have it in CLI.

There is no need to create a perfect solution here, but my worry is that with the current approach a command which worked before may get stuck now. 

I think that's correct behavior to retry domain occupied errors forever (if the goal is to process all requests). However, there is a difference from e.g. throttling errors, which also use stop_never. If there is a throttling error, there is no other work you can do in the meantime; throttling happens for all connections, and it'll resolve itself in some time. Domain occupied error is specific to a particular query, and you can do other work while waiting for next retry opportunity. Once a domain is occupied, it is easy to get all available connections stuck, having them retry this error just for a single domain.

Unfortunately this applies to other query errors as well. Network/throttling/server errors are not caused by a single website, they are overall system errors, while query errors do depend on the URL being processed. It seems my proposal of doing retrials inside request_raw was not that good in the end :(

If it works in practice, that's probably fine. A similar approach worked for Scrapy retrials for quite a long time; it was not perfect (sometimes spiders get stuck), but workable. Scrapy was not using infinite retries though - with infinite the behavior could have been much worse, with no chance to recover. So, to make domain occupied errors no worse than query errors, I wonder if we should make their retrials bounded as well for now.

What do you think @victor-torres?

Reply:
I appreciate your thoughts, Mikhail.

I don't like the "stop never" strategy because it's like having a "while true" condition. It could lead to starvation and waste of computational resources. There should be an exit condition. Even with Domain Occupied errors because you never know.

I'm not comfortable with the long thresholds too, having to wait, for example, 15 minutes before giving up. To some extent, I think the "fail fast" strategy could be applied in those cases. For example, we don't need to spend too much time in our spider waiting for failing queries. We could schedule a supplementary job to grab missing items later. The key here is handing responsibility back to the spider. It should be able to make the best decision: to wait for a longer period of time or to give up temporarily and try again later. Of course, we could supply a better API to make it possible, like the max retry count you've suggested in previous comments.

We knew beforehand that trying to implement such a retry strategy with that granularity and also supporting sub-batches would be a challenge. And it will probably never cover 100% of our use cases. Similar libraries have a much simpler retry strategy. I remember one thing that was discussed was that this retry strategy should be used with small batches, probably to minimize issues like the ones you described. Of course, this is something developers should have in mind while creating solutions using this library, and having to remember things or knowing implementation details is not an ideal situation. 

So trying to be a little bit pragmatic here, we have a couple of options to improve this solution without refactoring a significant part of it:

- we could replace the "stop never" strategy by a finite (and safer) alternative
- we could create a separate flag for Domain Occupied errors

There's also the option to use a max retry count, but I'll discuss it in another comment because I think it's a world apart.

Oh, almost forgot that we had internal discussions to remove the batch support from this wrapper since almost no one is making use of it. Will it be useful in the near future? Is it necessary to overcome some API limitations such as rate limits and RPS? Would the previous issues happen if single requests are made per user query?

Reply:
Domain occupy seems to be the kind of error that should be better handled by the user (e.g. he can decide to retry on it tomorrow). So I'm Ok to set a limit for it and return it back to the user so that It handles it the best way he can. 

Regarding batch, I think is generating more troubles that time it saves. We should reconsider supporting only single requests. It will be more maintainable and with have a better behaviour.

> Similar libraries have a much simpler retry strategy

The problem is that AutoExtract API as it is designed now have a considerable number of failed responses that requires a retrial. Leaving the retrial logic to the user makes the library simpler but at the same time it is less useful to the user given that he needs to spend considerable efforts to implement the retrial logic (or don't do it and get a suboptimal service quality).

Reply:
Leaving here for history.
I see it as two different things.
For a low level library, I expect it to run at most a few retries and return. In general, I don't expect to call a python function and it will loop forever, this is a kind of huge side effect.
On the other hand, for a CLI, I think it is fine to loop forever, as I want the results, no matter the time it takes to get them.
So, maybe a better separation of concerns of what is the AutoExtract integration library (wrapper to HTTP calls) and what is a CLI framework that can be used to extract data from a set of URLs.

Reply:
@victor-torres 

>  we could replace the "stop never" strategy by a finite (and safer) alternative
    we could create a separate flag for Domain Occupied errors

+1 to replace stop_never for now, and move ahead. I think we can leave without special logic for Domain Occupied. We can e.g. respect time provided in messages, but reuse the same max-retry amount.

> Oh, almost forgot that we had internal discussions to remove the batch support from this wrapper since almost no one is making use of it. Will it be useful in the near future? Is it necessary to overcome some API limitations such as rate limits and RPS? Would the previous issues happen if single requests are made per user query?

I think the problem described here is not related to batches, and won't be solved by removing batching support.
The issue is cause by the way retries are handled. Currently architecture looks like this (assuming concurrency is set to 3):

```
[List of URLs] 
   -> request_raw #1, retry until success 
   -> request_raw #2, retry until success 
   -> request_raw #3, retry until success          
```

An architecture which would avoid this issue:

```
[List of URLs] -> [Queue]
   -> request_raw #1, make a single request, if retry is needed, add a retry request to Queue
   -> request_raw #2, make a single request, if retry is needed, add a retry request to Queue
   -> request_raw #3, make a single request, if retry is needed, add a retry request to Queue
```

I think changing it is out of scope here, unless @victor-torres you see an easy way to do it.

Reply:
No, I think we can leave this change to a queue/scheduler for another opportunity.